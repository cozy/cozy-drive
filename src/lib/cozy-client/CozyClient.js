/* global cozy */
import DataAccessFacade from './DataAccessFacade'
import SharingsCollection, {
  SHARINGS_DOCTYPE
} from './collections/SharingsCollection'
import AppsCollection, { APPS_DOCTYPE } from './collections/AppsCollection'
import { authenticateWithCordova } from './authentication/mobile'
import { getIndexFields, isV2 as isV2Helper } from './helpers'

const FILES_DOCTYPE = 'io.cozy.files'

export default class CozyClient {
  constructor(config) {
    const { cozyURL, ...options } = config
    this.options = options
    this.collections = {}
    this.indexes = {}
    this.specialDirectories = {}
    this.facade = new DataAccessFacade()
    this.store = null
    if (cozyURL) {
      this.facade.setup(cozyURL, options)
    }
    this.defineSpecialCollections()
  }

  register(cozyUrl) {
    this.facade.setup(cozyUrl, {
      ...this.options,
      oauth: {
        ...this.options.oauth,
        onRegistered: (client, url) => authenticateWithCordova(url)
      }
    })
    return cozy.client.authorize(true)
  }

  attachStore(store) {
    this.store = store
  }

  resetStore() {
    this.store.dispatch({ type: 'RESET_STORE' })
    this.facade.destroyAllDatabases()
  }

  getUrl() {
    return this.facade.getUrl()
  }

  isV2(cozyURL) {
    return isV2Helper(cozyURL)
  }

  async isRegistered(clientInfos) {
    if (!clientInfos) return false
    try {
      await cozy.client.auth.getClient(clientInfos)
      return true
    } catch (err) {
      if (err.message === 'Client has been revoked') {
        return false
      } else {
        console.log(
          'Error while retrieving oauth client information, but client is not revoked'
        )
        console.warn(err)
        return true
      }
    }
  }

  defineCollection(doctype, collectionAdapter) {
    this.collections[doctype] = collectionAdapter
  }

  defineSpecialCollections() {
    this.defineCollection(SHARINGS_DOCTYPE, new SharingsCollection())
    this.defineCollection(APPS_DOCTYPE, new AppsCollection())
  }

  getCollection(doctype) {
    if (!this.collections[doctype]) {
      throw new Error(`No collection found for doctype ${doctype}`)
    }
    return this.collections[doctype]
  }

  startSync(dispatch) {
    return this.facade.startSync(dispatch)
  }

  startReplicationTo(dispatch) {
    return this.facade.startReplicationTo(dispatch)
  }

  startReplicationFrom(dispatch) {
    return this.facade.startReplicationFrom(dispatch)
  }

  getAdapter(doctype) {
    return this.facade.getAdapter(doctype)
  }

  async fetchDocuments(queryName, doctype, options = {}, skip = 0) {
    if (options.selector) {
      const index = await this.getQueryIndex(queryName, doctype, options)
      return this.getAdapter(doctype).queryDocuments(doctype, index, {
        ...options,
        skip
      })
    }
    return this.getAdapter(doctype).fetchDocuments(doctype)
  }

  fetchDocument(doctype, id) {
    return this.getAdapter(doctype).fetchDocument(doctype, id)
  }

  fetchFile(id) {
    return this.getAdapter(FILES_DOCTYPE).fetchFile(id)
  }

  // TODO: temp method for drive mobile so that we can fetch the thumbnails links
  // that are absent from the Pouch docs
  fetchFilesForLinks(folderId, skip = 0) {
    return this.facade.stackAdapter.fetchFilesForLinks(folderId, skip)
  }

  fetchReferencedFiles(doc, skip = 0) {
    return this.getAdapter(doc._type).fetchReferencedFiles(doc, skip)
  }

  addReferencedFiles(doc, ids) {
    return this.getAdapter(doc._type).addReferencedFiles(doc, ids)
  }

  removeReferencedFiles(doc, ids) {
    return this.getAdapter(doc._type).removeReferencedFiles(doc, ids)
  }

  createDocument(doctype, doc) {
    return this.getAdapter(doctype).createDocument(doctype, doc)
  }

  updateDocument(doc) {
    return this.getAdapter(doc._type).updateDocument(doc)
  }

  /**
   * Update documents in bulk.
   *
   * All documents matching the query will be retrieved before updating.
   *
   * @example
   * ```
   * await dispatch(
   *   updateDocuments(
   *     'io.cozy.bank.transactions',
   *     {
   *       selector: { accountId: '1921680010' }
   *     },
   *     {
   *       updateCollections: ['transactions']
   *     },
   *     transaction => ({ ...transaction, amount: transaction.amount + 10 })
   *   )
   * )
   * ```
   *
   * @param  {String} doctype  - Doctype of the documents that will be updated
   * @param  {Object} query    - Mango query to select which documents will be updated
   * @param  {Function} iterator - Function that will update the documents
   * @return {Promise}
   */
  updateDocuments(doctype, query, iterator) {
    return this.getAdapter(doctype).updateDocuments(doctype, query, iterator)
  }

  deleteDocument(doc) {
    return this.getAdapter(doc._type).deleteDocument(doc)
  }

  /**
   * Delete documents in bulk.
   *
   * All documents matching the query will be retrieved before deleting.
   *
   * @example
   * ```
   * await dispatch(deleteDocuments('io.cozy.bank.operations', {
   *   selector: { account: account.id }
   * }, {
   *   updateCollections: ['transactions']
   * }))
   * ```
   *
   */
  deleteDocuments(doctype, query) {
    return this.getAdapter(doctype).deleteDocuments(doctype, query)
  }

  createFile(file, dirID) {
    return this.getAdapter(FILES_DOCTYPE).createFile(file, dirID)
  }

  trashFile(file) {
    return this.getAdapter(FILES_DOCTYPE).trashFile(file)
  }

  async ensureDirectoryExists(path) {
    if (!this.specialDirectories[path]) {
      const dir = await cozy.client.files.createDirectoryByPath(path)
      this.specialDirectories[path] = dir._id
    }
    return this.specialDirectories[path]
  }

  async checkUniquenessOf(doctype, property, value) {
    const index = await this.getUniqueIndex(doctype, property)
    const existingDocs = await cozy.client.data.query(index, {
      selector: { [property]: value },
      fields: ['_id']
    })
    return existingDocs.length === 0
  }

  async getQueryIndex(queryName, doctype, options) {
    if (!this.indexes[queryName]) {
      this.indexes[queryName] = await this.getAdapter(doctype).createIndex(
        doctype,
        getIndexFields(options)
      )
    }
    return this.indexes[queryName]
  }

  async getUniqueIndex(doctype, property) {
    const name = `${doctype}/${property}`
    if (!this.indexes[name]) {
      this.indexes[name] = await this.getAdapter(doctype).createIndex(doctype, [
        property
      ])
    }
    return this.indexes[name]
  }
}
