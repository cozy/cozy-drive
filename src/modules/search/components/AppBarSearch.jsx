import React from 'react'

import { BarSearch } from 'cozy-bar'
import { AssistantDesktop } from 'cozy-dataproxy-lib'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'

const AppBarSearch = () => {
  const { isMobile } = useBreakpoints()

  return (
    <BarSearch>
      {!isMobile && (
        <AssistantDesktop
          componentsProps={{ SearchBarDesktop: { size: 'small' } }}
        />
      )}
    </BarSearch>
  )
}

export default AppBarSearch
