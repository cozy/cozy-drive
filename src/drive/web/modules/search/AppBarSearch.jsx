import React from 'react'

import { useI18n } from 'cozy-ui/transpiled/react/providers/I18n'
import useBreakpoints from 'cozy-ui/transpiled/react/providers/Breakpoints'

import { BarSearch } from 'components/Bar'
import BarSearchAutosuggest from 'drive/web/modules/search/components/BarSearchAutosuggest'

const AppBarSearch = () => {
  const { t } = useI18n()
  const { isMobile } = useBreakpoints()

  return <BarSearch>{!isMobile && <BarSearchAutosuggest t={t} />}</BarSearch>
}

export default AppBarSearch
