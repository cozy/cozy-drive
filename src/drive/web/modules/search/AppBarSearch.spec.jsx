import React from 'react'
import { render } from '@testing-library/react'

import CozyClient from 'cozy-client'

import AppBarSearch from 'drive/web/modules/search/AppBarSearch'
import AppLike from 'test/components/AppLike'

jest.mock('drive/web/modules/services/components/iconContext', () => ({
  getIconUrl: () => 'iconUrl'
}))

it('should display the Searchbar', () => {
  const client = new CozyClient({})
  const appBarSearch = render(
    <AppLike client={client}>
      <AppBarSearch />
    </AppLike>
  )
  expect(appBarSearch).toMatchSnapshot()
})
