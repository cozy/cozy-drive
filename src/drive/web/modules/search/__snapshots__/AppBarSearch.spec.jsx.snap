// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the Searchbar 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="u-flex u-flex-items-center u-flex-grow CozyTheme--normal"
      >
        <div
          class="bar-search-container"
          role="search"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-owns="react-autowhatever-1"
            class="u-w-100"
            role="combobox"
          >
            <div
              class="styles__c-inputgroup___12OVJ styles__c-inputgroup--fullwidth___3nuay bar-search-input-group"
            >
              <div
                class="styles__c-inputgroup-side___60v0v"
              >
                <svg
                  aria-hidden="true"
                  class="bar-search-input-group-append styles__icon___23x3R"
                  height="16"
                  viewBox="0 0 16 16"
                  width="16"
                >
                  <path
                    d="M10.472 10H11.5l3.294 3.294a.996.996 0 010 1.412l-.088.088a1 1 0 01-1.412 0L10 11.5v-1.028a6 6 0 11.472-.472zM6 10a4 4 0 100-8 4 4 0 000 8z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
              <div
                class="styles__c-inputgroup-main___1LP4B"
              >
                <input
                  aria-autocomplete="list"
                  aria-controls="react-autowhatever-1"
                  autocomplete="off"
                  placeholder="Search anything"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <ul
              class="MuiList-root bar-search-autosuggest-suggestions-container MuiList-padding"
              id="react-autowhatever-1"
              role="listbox"
            />
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="u-flex u-flex-items-center u-flex-grow CozyTheme--normal"
    >
      <div
        class="bar-search-container"
        role="search"
      >
        <div
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-owns="react-autowhatever-1"
          class="u-w-100"
          role="combobox"
        >
          <div
            class="styles__c-inputgroup___12OVJ styles__c-inputgroup--fullwidth___3nuay bar-search-input-group"
          >
            <div
              class="styles__c-inputgroup-side___60v0v"
            >
              <svg
                aria-hidden="true"
                class="bar-search-input-group-append styles__icon___23x3R"
                height="16"
                viewBox="0 0 16 16"
                width="16"
              >
                <path
                  d="M10.472 10H11.5l3.294 3.294a.996.996 0 010 1.412l-.088.088a1 1 0 01-1.412 0L10 11.5v-1.028a6 6 0 11.472-.472zM6 10a4 4 0 100-8 4 4 0 000 8z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <div
              class="styles__c-inputgroup-main___1LP4B"
            >
              <input
                aria-autocomplete="list"
                aria-controls="react-autowhatever-1"
                autocomplete="off"
                placeholder="Search anything"
                type="text"
                value=""
              />
            </div>
          </div>
          <ul
            class="MuiList-root bar-search-autosuggest-suggestions-container MuiList-padding"
            id="react-autowhatever-1"
            role="listbox"
          />
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
